tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+		"graceful-recovery/gateway.yaml",
+		"graceful-recovery/cafe-routes.yaml",
+	}
+	ns := &core.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "graceful-recovery",
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+	})
+
+	AfterAll(func() {
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+	})
+
+	It("recovers when NGF container is restarted", func() {
+		runRecoveryTest(ngfContainerName, files, ns)
+	})
+
+	It("recovers when nginx container is restarted", func() {
+		runRecoveryTest(nginxContainerName, files, ns)
+	})
+})
+
+func runRecoveryTest(containerName string, files []string, ns *core.Namespace) {
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	leaseName, err := getLeaderElectionLeaseHolderName()
+	Expect(err).ToNot(HaveOccurred())
+
+	restartContainer(containerName)
+
+	checkContainerLogsForErrors(podNames[0])
+
+	Expect(waitForLeaderLeaseToChange(leaseName)).ToNot(HaveOccurred())
+
+	Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+
+	Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+
+	Expect(waitForFailingTraffic()).ToNot(HaveOccurred())
+
+	Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+	Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+	Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+}
+
+func restartContainer(containerName string) {
+	var jobScript string
+	if containerName == "nginx" {
+		jobScript = "PID=$(pgrep -f \"[n]ginx: master process\") && kill -9 $PID"
+	} else {
+		jobScript = "PID=$(pgrep -f \"/[u]sr/bin/gateway\") && kill -9 $PID"
+	}
+
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	restartCount, err := getContainerRestartCount(containerName, podNames[0])
+	Expect(err).ToNot(HaveOccurred())
+
+	job, err := runNodeDebuggerJob(podNames[0], jobScript)
+	Expect(err).ToNot(HaveOccurred())
+
+	Expect(waitForContainerRestart(podNames[0], containerName, restartCount)).ToNot(HaveOccurred())
+
+	// propagation policy is set to delete underlying pod created through job
+	Expect(resourceManager.Delete(
+		[]client.Object{job},
+		client.PropagationPolicy(metav1.DeletePropagationBackground),
+	)).ToNot(HaveOccurred())
+}
+
+func waitForContainerRestart(ngfPodName, containerName string, currentRestartCount int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
```

Comment:
RequestTimeout sounds quite different from container restart timeout. I suggest having a separate timeout value here, so that changes to RequestTimeout don't affect this test.